primitives/fee-handler/src/fee_handler.rs

```
@@ -0,0 +1,370 @@
+
+		let remaining_credit2 =
+			Self::try_propagate_chain_fee(remaining_credit, payer, affiliate_id)?;
+
+		Self::deposit_into_treasury(treasury_pot, remaining_credit2)
+	}
+
+	fn withdraw_and_deposit_into_treasury(
+		who: &Self::AccountId,
+		asset_id: Self::AssetId,
+		treasury_pot: &Self::AccountId,
+		amount: Self::Balance,
+	) -> Result<(), DispatchError> {
+		let credit = W::withdraw_credit(who, asset_id, amount)?;
+		Self::deposit_into_treasury(treasury_pot, credit)
+	}
+}
+
+impl<AccountId, Assets, W, Affiliate, Tournament>
+	AssetGameFeeHandler<AccountId, Assets, W, Affiliate, Tournament>
+where
+	Assets: fungibles::Inspect<AccountId, Balance = W::Balance, AssetId = W::AssetId>
+		+ fungibles::Balanced<AccountId>,
+
+	W: WithdrawCredit<
+		AccountId = AccountId,
+		Assets = Assets,
+		Credit = fungibles::Credit<AccountId, Assets>,
+	>,
+
+	Affiliate: DistributeFee<AccountId = W::AccountId, Balance = W::Balance>,
+	Tournament:
+		DistributeFee<AccountId = AccountId, Balance = W::Balance, MaxDistributions = ConstU32<1>>,
+{
+	/// Distributes an already withdrawn `fee_credit` to the affiliates of `account`.
+	///
+	/// Returns the remaining `fee_credit` after this operation.
+	fn try_propagate_chain_fee(
+		fee_credit: W::Credit,
+		account: &W::AccountId,
+		identifier: &Affiliate::FeeIdentifier,
+	) -> Result<W::Credit, DispatchError> {
+		let mut final_fee = fee_credit;
+
+		if let Some(a) = Affiliate::distribute_fee(final_fee.peek(), account, identifier) {
+			for allocation in a {
+				if allocation.amount > 0_u32.into() {
+					let affiliate_fee = final_fee.extract(allocation.amount);
+
+					if let Err(credit) = W::Assets::resolve(&allocation.beneficiary, affiliate_fee)
+					{
+						log::error!(
+							"Could not deposit to affiliate account, it probably doesn't exist."
+						);
+
+						// Reabsorb the credit; it can still be used.
+						let _ = final_fee.subsume(credit).defensive();
+					}
```

Shouldn't we just let the method fail if any of the allocations cannot be distributed? They way it is implemented right now, it would be quite difficult to trace any error that happens in `resolve`

---

The reasoning here is, that there are only 3 ways that this can fail:

1. The asset ID doesn't exist (not applicable in our case, as this would fail earlier in our flow)
2. The total issuance would overflow (highly unlikely)
3. The beneficiary doesn't exist currently, and the amount to be deposited is not enough to create an account or the deposited asset is not a sufficient (aka doesn't have the privileges to create a new account).

The problem is, if one of your affiliates is reaped because he has decided to send all his funds somewhere. He would break the transactions involving fees for another account, and I think this is a highly unfavorable behavior.

The same is true for depositing to the tournament treasury, but here I think it is better to skip allocating to those accounts instead of breaking transactions. Maybe I can elaborate this in the documentation.

---

Hmm, I understand what you mean, yeah I think with proper documentation that should be clear enough